src/Memory/MemoryInfoProducer.cpp

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "Memory/MemoryInfoProducer.h"
+
+#include <chrono>
+#include <thread>
+
+#include "MemoryUtils.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Profiling.h"
+#include "capture.pb.h"
+
+namespace orbit_memory {
+
+using orbit_grpc_protos::TotalMemoryUsageInfo;
+
+void MemoryInfoProducer::Start() {
+  *exit_requested_ = false;
+  thread_ = std::make_shared<std::thread>(&MemoryInfoProducer::Run, this);
+}
+
+void MemoryInfoProducer::Stop() {
+  *exit_requested_ = true;
+  if (thread_ != nullptr && thread_->joinable()) {
+    thread_->join();
+  }
+  thread_.reset();
+}
+
+void MemoryInfoProducer::Run() {
+  pthread_setname_np(pthread_self(), "MemoryInfoProducer::Run");
+
+  CHECK(listener_ != nullptr);
+  ProduceTotalMemoryUsageInfoAndSendToListener();
+}
+
+void MemoryInfoProducer::ProduceTotalMemoryUsageInfoAndSendToListener() {
+  std::optional<TotalMemoryUsageInfo> total_memory_usage_info;
+  uint64_t current_timestamp_ns = orbit_base::CaptureTimestampNs();
+  while (!(*exit_requested_)) {
+    total_memory_usage_info = GetTotalMemoryUsageInfo(current_timestamp_ns);
+    if (total_memory_usage_info.has_value()) {
+      listener_->OnTotalMemoryUsageInfo(total_memory_usage_info.value());
+    }
+
+    std::this_thread::sleep_for(std::chrono::nanoseconds(sampling_period_ns_));
```

Comment:
I think that the sampling period should be handled differently. Only sleeping for `sampling_period_ns_` doesn't account for how long the sleep will actually be (this value passed is just a lower bound) and for how long the rest of the code in the `while` will take. This means that over time the sampling period will be higher than the one passed. An approach similar to what games do to compute how much time is left before the next frame should probably be used.

Reply:
Thanks for pointing out this and providing the helpful suggestions!

I changed to compute the scheduled running time for the next memory sampling (i.e., current scheduled time + sampling period), and use the `sleep_until` method to block the thread until the scheduled running time.

I'm not sure about how this has been done in games, but I take a look at the code in Perfetto. 
After finishing current round of memory sampling, they create a delayed task for the next round of memory sampling.
The delayed task includes the memory sampling function and the scheduled time for running.  They use a `UnixTaskRunner` to manage all delayed tasks in a `std::multimap<TimeMillis, std::function<void()>> delayed_tasks_`. This `UnixTaskRunner` will keep comparing the scheduled time of the delayed tasks with the current time to determine whether this task should be run and erased.

Reply:
This makes sense to me, thanks.